GitCommands/Git/GitModule.cs

```diff
@@ -3859,5 +3858,15 @@ public bool StopTrackingFile(string filename)
         {
             return RunGitCmdResult("rm --cached " + filename).ExitedSuccessfully;
         }
+
+        /// <summary>
+        /// Determines whether a git command's output indicates an error occurred.
+        /// </summary>
+        /// <param name="gitOutput">The output from the git command, to inspect.</param>
+        /// <returns><c>true</c> if the command detailed an error, otherwise <c>false</c>.</returns>
+        public static bool IsGitErrorMessage(string gitOutput)
+        {
+            return Regex.IsMatch(gitOutput, @"^\s*(error:|fatal)");
```

Comment:
nit: it's more efficient to have static readonly instances of Regex creadted with .Compiled option whenever possible

Reply:
It's times like this I wish C# had static locals, like C++ does. Seems a shame to introduce a type-level field for something that's only used in one member. We now have nested functions. Why not nested static fields too?

I'd be interested in some benchmark results here, as well as info on runtime compilation times (consider it might cache `DynamicMethod` internally or similar). I suspect the difference is negligible, especially for this regex pattern with little seeking about. And there's a trade-off: the compilation of such static Regex objects would happen when their containing type first loads, meaning more latency when GE starts up.

Reply:
In this case the wins are probably low since I guess this method is called rarely but in a tight loop this would cause a measurable GC pressure since it appears each static call creates 208B of objects to clean up.

![image](https://user-images.githubusercontent.com/483659/37529755-94ce6f0c-2940-11e8-984f-6d45fb275585.png)

Reply:
Thanks for following up. Given we're talking nanos (not even micros) I don't think there's much to do here. Interesting to see the difference all the same. Allocation stats are interesting too. Will keep that in mind for future.

Are there any tight loops in GE? I can't think of any off the top of my head from what I've seen so far.

Reply:
Yeah.. I was glad and disappointed at the same time that execution took only nanos :D CPU perf wise this is absolutely not an issue, so only mem might be of a concern (not in this case though).
At some future time I'll do some benchmarking with proper tools to see where actual perf-hogs are.

Reply:
> It's times like this I wish C# had static locals, like C++ does. Seems a shame to introduce a type-level field for something that's only used in one member. 

I question why is this method part of `GitModule` in the first place. 
Is there any thing stopping us from creating something like the following:

```
public class GitCommandOutputTester : IGitCommandOutputTester 
{
    private static readonly Regex regex = ... // compiled

    public bool IsError(string commandOutput)
    {
        return regex.IsMatch(gitOutput, @"^\s*(error:|fatal)");
    }
}
```

Reply:
@mdonatas: what is this tool you use?

Reply:
> I question why is this method part of GitModule in the first place.

Fair question. In this case it's only because that's where the bulk of this stuff lives currently. With the approach laid out in #4643 `IsGitErrorMessage` would probably become part of the contract/documentation/implementation of `IGitExecutable`, and probably wouldn't even be a method but rather an inline test followed by a throw/return.

> Is there any thing stopping us from creating...

Of course nothing is stopping that. For my taste that level of granularity for interface behaviour is too fine, especially when there would only be one implementation for the foreseeable future. Sometimes adding this abstraction makes total sense, but there are thousands of small bits of policy in the handling of git operations (think `GitModule`), and if we create interfaces/implementations/injections/lifecycles for all of them I think the result will be confusing and overwhelming. I know that it is extreme to suggest all policies would need interfaces, and I don't mean it to sound dramatic but rather illustrative of the fact there's a level of taste/discretion on such a decision.

Thinking about it some more, I think a "good design" for solution to a problem looks simple and obvious at a coarse level, with fine level abstractions added to support flexibility where needed. Over the course of a monolith's life, usually fine abstractions accumulate to deal with the fact that the coarse abstractions no longer quite meet the changed requirements/usage. While we're talking about breaking up the monolith of `GitModule`, I vote for thinking/talking mostly about these coarse abstractions, as the fine ones can be added/tweaked comparatively easily after the fact. Amending coarse abstractions is a much harder form of debt to pay off.

That's been my experience anyway.

Reply:
Drew, I agree in principal. 
However working with legacy projects (likes of GE) one isn't always in a position we are now - where we can re-imagine and re-work the structure. We are also extremely lucky to have you 🎆 
Quite often one is tasked with specific problems which must be addressed within allocated time. In such cases chipping off bits of monolith into fine grained SOLID classes (which then can be tested and reasoned about) is usually the only viable approach.

I've also been writing frameworks and shared assets for some years now and I'm used to breaking problems down to fine abstractions.

Reply:
RussKie, sorry I missed your comment.

Dealing with any code base is like playing chess. There is always a move :) One must think both short and long term.

Now that I've spent a bit of time with GE I feel like my first proper contribution will be around consolidating git process execution, and the strategy for `GitModule` refactoring (#4643). I see a series of moves that get us from where we are now to a point where the code is clearer, with better abstractions and is more testable. The design and coding is the easy part :) The effort on my part is to communicate effectively and ensure the solution meets the goals of the team, including the goals I don't yet see.

I have a working implementation of the git process execution code and would like to get it in front of you all. It's blocked by the few open PRs here (#4641, #4639 and especially #4662), so I want to do whatever I can to get these merged or closed and move forward.

Reply:
I'll try to get to them as soon as can.

Reply:
> The effort on my part is to communicate effectively and ensure the solution meets the goals of the team, including the goals I don't yet see.

:+1: 
There are few things which are important to me in any project, applicable to both UI and API - consistency, usability, readability, testability and maintainability. These are interconnected and compliment one another.
SOLID principles seem to help a lot in achieving the above, hence I'm so obsessed with them :)

Reply:
> Dealing with any code base is like playing chess. There is always a move :) One must think both short and long term.

Yes, however OS has its caveats - if at (paid) work one has a greater degree of commitment and hence can afford greater degree of trust in promises of future delivery, in the OS world the degree of commitment is lower - family, paid work and other real life commitments have higher priorities (understandably so). People also lose interest...
In this case a promise of future refactor or rework may never be fulfilled, and the codebase would left with a rot.

Reply:
> People also lose interest...

I've run open source projects for a long while and know how it goes. GitHub is filled with well intentioned half-promises to look into things and, well, life gets in the way or it turns out to be harder than expected, or whatever.

In my case I hope you can tell that I'm at least a bit invested in and committed to this project. How long that will last, I don't know. But for now I have energy and time.

But I don't think that matters. These PRs all improve the code along at least one of the dimensions you mentioned, leaving things in better shape than they were beforehand. I don't see any unfulfilled promise that would lead to rot, but maybe I suffer from a biased perspective.

GE isn't very testable right now. I'm happy to help kick things in that direction.

Oh and thanks for your kind words earlier. GE is also extremely lucky to have someone as committed as you taking the lead and driving things forward. I know how much time it takes.